### DISCUSSION TITLE
Node and Async

### DISCUSSION THREAD

**Discussion by elyobo** [duckdb/duckdb#5619] [Ideas]
We currently use SQLite for some projects and it seems like DuckDB might be a more appropriate tool for some. With these projects we make use of [`better-sqlite`](https://github.com/WiseLibs/better-sqlite3) instead of the `node-sqlite3` lib, as we're reasonably convinced [by the rationale from the `better-sqlite` authors](https://github.com/WiseLibs/better-sqlite3#why-should-i-use-this-instead-of-node-sqlite3).

I notice that the [provided duckdb nodejs API](https://duckdb.org/docs/api/nodejs/overview) is async (and callback based!); does the rationale for a sync API for SQLite hold up for DuckDB as well?

**Comment by Mause** [comment]
That sort of approach isn't really possible in DuckDB for the following reason (taken from the better-sqlite3 readme)

> For example, if you're executing queries that take one second to complete, and you expect to have many concurrent users executing those queries, no amount of asynchronicity will save you from SQLite3's serialized nature.

**Reply by elyobo** [reply]
You mean that internally DuckDB isn't serialised like SQLite is so this isn't a problem (and so async access is still useful)?

**Reply by Mause** [reply]
> You mean that internally DuckDB isn't serialised like SQLite is so this isn't a problem (and so async access is still useful)?

I mean that when queries take over, say, 500ms to complete, as is often the case with the scale of data that people are usually analysing with DuckDB, the NodeJS will be blocked for that period of time, and unable to do any other computation

**Reply by elyobo** [reply]
My (mis?)understanding in the SQLite case was that this is the same there too, but even in async node still can't do anything because SQLite already is and it's in the same single threaded process - i.e. async doesn't solve this problem for SQLite, slow queries block you sync or async.

Am I misunderstanding that or is it that DuckDB is different here.

**Reply by Mause** [reply]
See my reply to https://github.com/duckdb/duckdb/discussions/5619#discussioncomment-4386489, I've tried to explain it a bit more there

**Reply by elyobo** [reply]
Saw that and yes, that answers my question perfectly. DuckDB is doing work in another thread and so does allow the main thread to continue work, meaning that there is a real advantage to the async API here.

**Reply by elyobo** [reply]
I also notice that there's a [promise based API](npmjs.com/duckdb-async) coming as well :tada:

**Comment by axdg** [comment]
>I mean that when queries take over, say, 500ms to complete, as is often the case with the scale of data that people are usually analysing with DuckDB, the NodeJS will be blocked for that period of time, and unable to do any other computation

I'm having trouble understanding this (in relation to the OPs question)... 

If any call from NodeJS to DuckDB (such as executing a query that takes 500ms) is going to block the NodeJS process until results are returned, then (I think) there isn't going to be any benefit to using a callback / `async` / `Promise` based API (as opposed to a synchronous one), right?

**Reply by Mause** [reply]
You're almost correct - synchronous apis generally do the work in the same thread as the calling code, and are hence blocked until it completes. NodeJS is historically single threaded (and code with multiple threads is quite rare). The way the DuckDB bindings (and most other NodeJS bindings) avoid this, is doing the CPU and IO bound tasks in another thread, so as to not block the main thread. When the work completes, it adds an item to the NodeJS event loop, letting it know it can continue that piece of work. In the interim, other items on the event loop in the main thread can be processed

Hope this helps, let me know if you need any more explanation/if there's something wrong with it 😁

**Reply by elyobo** [reply]
Thanks Mause (and axdg for asking this question better than I did!), that makes perfect sense - in that context it's pretty clear that an async API does make sense.
